File: pointer_array_structure-master/sort_colors.py
class Solution:
    """
    @param nums: A list of integer which is 0, 1 or 2
    @return: nothing
    """

    def sortColors(self, nums):
        # write your code here
        # firstly, red -- blue then white--blue

        left, right = 0, len(nums) - 1

        while left < right:
            while left < right and nums[left] == 0:
                left += 1

            while left < right and nums[right] != 0:
                right -= 1

            if left < right:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right -= 1

        left, right = left, len(nums) - 1

        while left < right:
            while left < right and nums[left] == 1 or nums[left] == 0:
                left += 1

            while left < right and nums[right] == 2:
                right -= 1

            if left < right:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right -= 1

File: src/history/two-pointer/sort-colors.java
class Solution {
    public void sortColors(int[] arr) {
        int left = 0, right = arr.length - 1;
        int cur = 0;
        while(cur <= right){
          if(arr[cur] == 0){
            swap(arr, cur, left);
            cur += 1;
            left += 1;
          }
          else if(arr[cur] == 1){
            cur += 1;
          }
          else if(arr[cur] == 2){
            swap(arr, cur, right);
            right -= 1;
          }
        }
  }

  private void swap(int[] arr, int i, int j){
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }
}